### PR TITLE
npm2nix: 6.0.0

### PR DESCRIPTION
**Pull request by gilligan**
This is a first PR to work towards what was discussed in  #14532 
- Add npm2nix package (using https://github.com/svanderburg/npm2nix/tree/reengineering2):

The reengineering branch by @svanderburg is a big improvement on `NixOS/npm2nix` version. I hope you are fine with me packaging it.

/cc @zimbatm 

**Edit**: I first tried to actually remove npm2nix from node-packages.json / node-packages-generated not realizing that the new npm2nix version generates output that is not compatible with the previous one. I amended it to only add the package for now.

### CODE REVIEW

**Comment by gilligan** [comment]
```
error: getting status of ‘/home/travis/build/NixOS/nixpkgs/pkgs/top-level/test.nix’: No such file or directory
```

Did I break something? I have no idea what that is about x__X

**Comment by puffnfresh** [comment]
@gilligan yes, you modified pkgs/top-level/default.nix which I don't think you want to do.

**Comment by puffnfresh** [comment]
Otherwise :+1:

**Comment by gilligan** [comment]
@puffnfresh eeek. I screwed up somehow.. fix incoming shortly haha :)

**Comment by grahamc** [comment]
Not sure about these diff stats:
```
48,922 additions, 50,841 deletions not shown because the diff is too large.
```

**Comment by gilligan** [comment]
Sorry everyone.. i somehow accidentally replaced the entire  `pkgs/top-level/default.nix` with some nonsense ... FIxed now.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -6450,13 +6452,13 @@ in
   beecrypt = callPackage ../development/libraries/beecrypt { };
 
   beignet = callPackage ../development/libraries/beignet {
-    inherit (llvmPackages) clang-unwrapped; 
-    inherit (xlibs) libX11; 
-    inherit (xorg) libXfixes libpthreadstubs libXdmcp libXdamage libXxf86vm; 
-    inherit (python3Packages) python; 
-    inherit (purePackages) gl; 
-  }; 
-
```
Comment (zimbatm) [inline pkgs/top-level/all-packages.nix]: can you drop these whitespace changes ? not against it but they are unrelated to the PR

File: pkgs/development/tools/npm2nix/default.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    description = "Generate nix expressions to build npm packages";
+    homepage = http://github.com/svanderburg/npm2nix/tree/reengineering2;
+    platforms = with platforms; linux ++ darwin;
+  };
+
+  buildInputs = [ nodejs ];
+
+  buildPhase = ''
+    mkdir .npm-home
+    export HOME=$(readlink -f ./npm-home)
```
Comment (zimbatm) [inline pkgs/development/tools/npm2nix/default.nix]: Does `npm-home` exist ? The one just above is `.npm-home` with a dot in front.

**Comment by zimbatm** [comment]
Does npm2nix itself need to be inspected as well ?

According to our conventions I think the commit should be named `npm2nix: init at 6.0` since it's a new package.

**Comment by gilligan** [comment]
> Does npm2nix itself need to be inspected as well ?

I am hoping to hear from @svanderburg that he is OK with this.

File: pkgs/development/tools/npm2nix/default.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    rev = "758b286acfb077051a17aa9579adc558deeebbd4";
+    sha256 = "0gmvpxvp75fqw7w7x0c1ywc1zghd1hl6iimbv4xnjwlxgh704yg6";
+  };
+
+  meta = with lib; {
+    description = "Generate nix expressions to build npm packages";
+    homepage = http://github.com/svanderburg/npm2nix/tree/reengineering2;
+    platforms = with platforms; linux ++ darwin;
+  };
```
Comment (FRidh) [inline pkgs/development/tools/npm2nix/default.nix]: add yourself perhaps as maintainer?

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -2195,7 +2195,9 @@ in
   else
     nodePackages_4_x;
 
-  npm2nix = nodePackages.npm2nix;
+  npm2nix = callPackage ../development/tools/npm2nix {
+    inherit fetchFromGitHub;
```
Comment (joachifm) [inline pkgs/top-level/all-packages.nix]: Is `inherit` actually required here? I'd think `callPackage` would be able to fill it in.

**Comment by offlinehacker** [comment]
There's this alternative package manager for nodejs called ied https://github.com/alexanderGugel/ied that took ideas explicitly from nix, that would be really simple to generate nix expressions from. I would suggest to rather make a simple solution based on ied, as continue to have complex solution as is now. And as far as i looked, it's doable, but of course somebody has to do it.

**Comment by gilligan** [comment]
ied is not even close to be feature complete and thus sadly not a feasible replacement for now - plus considerable effort has been invested into npm2nix which is a working solution which shouldn't just be discarded.

My 2 cents ;)

**Comment by zimbatm** [comment]
@gilligan what is the license of npm2nix ? I see a warning emitted by npm that the field is missing.

**Comment by zimbatm** [comment]
`npm install` is trying to access the network so it won't work with the sandboxed builds. not sure why nox-review hasn't caught that.

**Comment by joachifm** [comment]
@zimbatm I think travis uses unsandboxed builds.

**Comment by garbas** [comment]
@gilligan @svanderburg is this PR obsolete?

**Comment by gilligan** [comment]
Oh, yes indeed

**Pull request closed without merging**